Alloy/MTLDevice+Extensions.swift

```diff
@@ -0,0 +1,53 @@
+//
+//  MTLDevice+Extensions.swift
+//  Alloy-iOS
+//
+//  Created by Vladimir Pavlov on 06/05/2019.
+//
+
+import Metal
+import CoreGraphics
+
+@available(iOS 11.0, macOS 10.12, *)
+public extension MTLDevice {
+
+    func maxTextureSize(desiredSize: CGSize) -> CGSize {
+        let maxSide: CGFloat
+        if self.supportsOnly8K() {
+            maxSide = 8192 * metalSafeMultiplier
+        } else {
+            maxSide = 16_384 * metalSafeMultiplier
+        }
+
+        guard desiredSize.width > 0, desiredSize.height > 0 else { return .zero }
+        let aspectRatio = desiredSize.width / desiredSize.height
+        if aspectRatio > 1 {
+            let resultWidth = min(desiredSize.width, maxSide)
+            let resultHeight = resultWidth / aspectRatio
+            return CGSize(width: resultWidth, height: resultHeight)
+        } else {
+            let resultHeight = min(desiredSize.height, maxSide)
+            let resultWidth = resultHeight * aspectRatio
+            return CGSize(width: resultWidth, height: resultHeight)
+        }
+    }
+
+    private func supportsOnly8K() -> Bool {
+        #if os(macOS)
+        return false
+        #else
+        if self.supportsFeatureSet(.iOS_GPUFamily1_v4) { return true }
+        if self.supportsFeatureSet(.iOS_GPUFamily2_v4) { return true }
+        if #available(iOS 12.0, *) {
+            if self.supportsFeatureSet(.iOS_GPUFamily1_v5) { return true }
```

Comment:
i think that's redundant

Reply:
Why? E.g., iPhone 5S and iPhone 6 support iOS 12.

Reply:
yup, but that's cumulative. if the device supporst GPUFamily1_v5 it automatically supports GPUFamily1_v[N] N < 5

we discussed that in person I think